### PR TITLE
[release/4.x] Backport of the refactor done to enable JS interpreter re-use, but without the feature itself

### PR DESCRIPTION
**Pull request by achamayou**
Backports the following commits to `release/4.x`:
 - [Allow JS interpreters (and global state) to be reused (#5564)](https://github.com/microsoft/CCF/pull/5564)
 
This also contains a cherry-pick of #5269, and changes to disable the feature. The goal of this backport is to enable further backports that contain bug fixes to the JS runtime.

### CODE REVIEW

**Comment by eddyashton** [comment]
@achamayou Looks like this is missing the CHANGELOG entry? That was initially in the wrong section, and then moved, so likely didn't merge correctly. The current release notes for [`5.0.0-dev1`](https://github.com/microsoft/CCF/releases/tag/ccf-5.0.0-dev1) describe this as:

```
- Added support for reusing JS interpreters, persisting global state. See [docs](https://microsoft.github.io/CCF/main/build_apps/js_app_bundle.html#reusing-interpreters) for more detail.

- **Note**: Full support for this requires an update to the `set_js_app` action in the constitution. The modification to the sample constitution is visible [here](samples/constitutions/default/actions.js).
```

**Comment by achamayou** [comment]
> @achamayou Looks like this is missing the CHANGELOG entry? That was initially in the wrong section, and then moved, so likely didn't merge correctly. The current release notes for [`5.0.0-dev1`](https://github.com/microsoft/CCF/releases/tag/ccf-5.0.0-dev1) describe this as:
> 
> ```
> - Added support for reusing JS interpreters, persisting global state. See [docs](https://microsoft.github.io/CCF/main/build_apps/js_app_bundle.html#reusing-interpreters) for more detail.
> 
> - **Note**: Full support for this requires an update to the `set_js_app` action in the constitution. The modification to the sample constitution is visible [here](samples/constitutions/default/actions.js).
> ```
> 
> EDIT - That "here" link isn't very helpful, should really point to the diff of `actions.js`. For this branch, that's something like this: https://github.com/microsoft/CCF/pull/5739/files#diff-6ece832da944809eec526d650430a0402251eaad86e38fe67254a4c76a9fba3d

Yes but I am busy removing the change first, one sec.

**Comment by achamayou** [comment]
@eddyashton ok, I think that now picks up the refactoring without enabling the feature. I've left the KV map and its documentation, because that seems harmless.

File: src/js/interpreter_cache.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      size_t freshness_marker) override
+    {
+      if (access != js::TxAccess::APP)
+      {
+        throw std::logic_error(
+          "JS interpreter reuse lru is only supported for APP "
+          "interpreters");
+      }
+
+      // Return a fresh interpreter, not stored in the cache
```
Comment (eddyashton) [inline src/js/interpreter_cache.h:34]: I think this deserves a comment to clarify that we've backported this architectural change, but on `4.x` we're disabling the behaviour _exactly here_. Also worth highlighting that this is a modified merge, not a direct cherry-pick, in the initial PR description (maybe even the title).

File: include/ccf/endpoint.h [hunk 2]
```diff
@@ -83,6 +83,23 @@ namespace ccf::endpoints
...
+
+    std::string key;
+
+    bool operator==(const InterpreterReusePolicy&) const = default;
+  };
+
+  void to_json(nlohmann::json& j, const InterpreterReusePolicy& grp);
+  void from_json(const nlohmann::json& j, InterpreterReusePolicy& grp);
+  std::string schema_name(const InterpreterReusePolicy*);
+  void fill_json_schema(nlohmann::json& schema, const InterpreterReusePolicy*);
```
[the remark this answers is not in the document]
Comment (achamayou) [reply include/ccf/endpoint.h:101]: I don't think these changes are risky, and they bring us closer to 5.x, which makes further patching easier, so I will keep them.

**Pull request merged**